File: 278-first-bad-version/278-first-bad-version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        let start = 0;
        let end = n - 1;
        
        while(start <= end){
            let middle = Math.floor(start + ( end - start) / 2);
            if(isBadVersion(middle)){
                end = middle - 1;
            }else{
                start = middle + 1;
            }
        }
        return start;
    };
};

File: 278-first-bad-version/278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 0;
        int end = n - 1;
        
        while(start <= end){
            int middle = start + (end - start) / 2;
            if(isBadVersion(middle)){
                end = middle - 1;
            }else{
                start = middle + 1;
            }
        }
        return start;
    }
}